### ISSUE TITLE
Meter and Statistic Unavailable

### ISSUE THREAD

**Issue by zackaryssmith** [kdeyev/eyeonwater#107]
The integration is unable to pull usage data from the meter, other sensors pull in fine. I am able to see all data from the eyeonwater sources directly. 

Logger: homeassistant
Source: components/recorder/statistics.py:2230
First occurred: 9:49:21 AM (3 occurrences)
Last logged: 9:58:32 AM
Error doing job: Task exception was never retrieved

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 291, in async_refresh
    await self._async_refresh(log_failures=True)
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 414, in _async_refresh
    self.async_update_listeners()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 168, in async_update_listeners
    update_callback()
  File "/config/custom_components/eyeonwater/sensor.py", line 130, in _state_update
    self.import_historical_data()
  File "/config/custom_components/eyeonwater/sensor.py", line 157, in import_historical_data
    async_import_statistics(self.hass, metadata, statistics)
  File "/usr/src/homeassistant/homeassistant/components/recorder/statistics.py", line 2230, in async_import_statistics
    raise HomeAssistantError("Invalid statistic_id")
homeassistant.exceptions.HomeAssistantError: Invalid statistic_id

Logger: homeassistant.components.sensor
Source: helpers/entity_platform.py:580
integration: Sensor (documentation, issues)
First occurred: 9:49:21 AM (3 occurrences)
Last logged: 9:58:31 AM
Error adding entity sensor.water_meter_REDACTED_temperature for domain sensor with platform eyeonwater

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 580, in _async_add_entities
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 892, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1359, in add_to_platform_finish
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1009, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1132, in _async_write_ha_state
    state, attr, capabilities, shadowed_attr = self.__async_calculate_state()
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1067, in __async_calculate_state
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1015, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 543, in state
    value = self.native_value
            ^^^^^^^^^^^^^^^^^
  File "/config/custom_components/eyeonwater/sensor.py", line 188, in native_value
    return self.meter.meter_info.sensors.endpoint_temperature.seven_day_min
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'endpoint_temperature'

![image](https://github.com/kdeyev/eyeonwater/assets/9200166/bc92c0f5-95fc-42da-8f5f-cc590399c59d)

**Comment by kdeyev** [comment]
@zackaryssmith thank you for reporting the issue.
Is it a new issue? or you just installed the extension for the first time?

**Comment by zackaryssmith** [comment]
@kdeyev I had it previously installed but had removed it due to the usage resulting in a negative number. This issue is appearing from a reconfiguration of the integration. 

Odd behavior, if I change the unit for the stats sensor I am able to get a number but it doesn't seem to have updated in the last week.

**Comment by kdeyev** [comment]
@zackaryssmith Could you please try to downgrade to extension version v2.5.6?
I think it may be a regression in the pyonwater module

**Comment by zackaryssmith** [comment]
@kdeyev I spent some time downgrading versions to see when it last worked and it seems that for my issue it was 2.5.0. Curiously, this is the update before the Statistics sensor was added. I still need to toggle the units to get any usage to show and the history import service does not work but it never has in my case. 

I also noticed that in the energy dashboard configuration, I have the following errors with the current versions:
![image](https://github.com/kdeyev/eyeonwater/assets/9200166/300c8b9f-2ac3-4896-9985-a95d2c9e5fa7)
![image](https://github.com/kdeyev/eyeonwater/assets/9200166/17097bde-5370-4e18-8410-c829abd9158e)

**Comment by kdeyev** [comment]
The warning messages are expected: there is short explanation in the extension wiki.
But I'm puzzled with the error messages you shared, my best guess there is a problem with the meter ID parsing. If you have some programming experience - it would be very helpful if could run the [pyonwater example](https://github.com/kdeyev/pyonwater/blob/main/example.py) with your credentials.
Thank you!

**Comment by kdeyev** [comment]
@zackaryssmith I cannot spot any problem: the output looks as expected.
I still suspect that it's may be something odd with the meter UUID:
```
homeassistant.exceptions.HomeAssistantError: Invalid statistic_id
```
Could you please check if it includes spaces or any special symbols? My meter UUID looks like: '200010106', and the meter sensor id: 'sensor.water_meter_200010106', statistic id: 'water_meter_200010106_statistic'.

**Comment by zackaryssmith** [comment]
@kdeyev When I run the example.py, I get a meter_uuid='19 digit number'.

 My water meter name in home assistant looks like "Water Meter 200010106+2019-07-15" with a entry ID of "sensor.water_meter_200010106_2019_07_15_2 "

The statistic entry name is "Water Meter 200010106+2019-07-15" with a entity ID of "sensor.water_meter_200010106_2019_07_15"

**Comment by kdeyev** [comment]
@zackaryssmith It seems that HA takes care of the special symbols. I can try to normalize the UUID by replacing '+' and '-' but '_', but I'm not really sure that it's the problem.
BTW, for energy dashboard you need to use an entity with '_statistic' suffix.

**Comment by zackaryssmith** [comment]
@kdeyev No worries, seems like it is specific to my meter/ install for now. Thank you for taking the time to look into it, I appreciate it.

**Comment by kdeyev** [comment]
@zackaryssmith please try this one: https://github.com/kdeyev/eyeonwater/releases/tag/v2.5.8-beta.4
It's prerelease version, so you may need to check "show beta" checkbox: https://hacs.xyz/docs/faq/select_version/

**Comment by zackaryssmith** [comment]
@kdeyev No luck. Statistic sensor still shows as unavailable. 

From the debug log when reloading the service:

2024-06-10 16:34:44.598 DEBUG (MainThread) [custom_components.eyeonwater.coordinator] Discovered 1 meter(s)
2024-06-10 16:34:44.599 DEBUG (MainThread) [custom_components.eyeonwater] Fetching latest data
2024-06-10 16:34:44.603 DEBUG (MainThread) [custom_components.eyeonwater.statistic_helper] last_stats {}
2024-06-10 16:34:44.604 ERROR (MainThread) [homeassistant.components.sensor] Error adding entity sensor.water_meter_123456_2019_07_15_temperature for domain sensor with platform eyeonwater
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 600, in _async_add_entities
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 914, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1355, in add_to_platform_finish
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1000, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1126, in _async_write_ha_state
    state, attr, capabilities, shadowed_attr = self.__async_calculate_state()
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1061, in __async_calculate_state
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1006, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 543, in state
    value = self.native_value
            ^^^^^^^^^^^^^^^^^
  File "/config/custom_components/eyeonwater/sensor.py", line 194, in native_value
    return self.meter.meter_info.sensors.endpoint_temperature.seven_day_min
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'endpoint_temperature'
2024-06-10 16:34:45.234 DEBUG (MainThread) [custom_components.eyeonwater] Finished fetching EyeOnWater data in 0.635 seconds (success: True)
2024-06-10 16:34:45.235 DEBUG (MainThread) [custom_components.eyeonwater.statistic_helper] last_imported_time None - data 2024-06-10 10:00:00-07:00
2024-06-10 16:34:45.235 INFO (MainThread) [custom_components.eyeonwater.statistic_helper] 59 data points found
2024-06-10 16:34:45.235 INFO (MainThread) [custom_components.eyeonwater.sensor] 59 data points will be imported
2024-06-10 16:34:45.235 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved (None)
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 288, in async_refresh
    await self._async_refresh(log_failures=True)
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 411, in _async_refresh
    self.async_update_listeners()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 165, in async_update_listeners
    update_callback()
  File "/config/custom_components/eyeonwater/sensor.py", line 133, in _state_update
    self.import_historical_data()
  File "/config/custom_components/eyeonwater/sensor.py", line 160, in import_historical_data
    async_import_statistics(self.hass, metadata, statistics)
  File "/usr/src/homeassistant/homeassistant/components/recorder/statistics.py", line 2235, in async_import_statistics
    raise HomeAssistantError("Invalid statistic_id")
homeassistant.exceptions.HomeAssistantError: Invalid statistic_id


From the System Logs:

**Comment by hrana** [comment]
Popping in here to state that my meter is named in the same format and I am having the same issue with custom integration. Let me know if there is a faster way to help debug and any additional info you require.

**Comment by kdeyev** [comment]
@zackaryssmith sorry for the delay.
@hrana welcome to the thread. 
I've looked how HA core verifies the ids and added similar check in the integration:
```
    if not valid_entity_id(statistic_id):
        raise Exception("Invalid statistic_id {statistic_id} for meter {meter.meter_id}")
```

Please try this one: https://github.com/kdeyev/eyeonwater/releases/tag/v2.5.8-beta.7

**Comment by zackaryssmith** [comment]
@kdeyev It is pulling data when on beta 7. 

I get this in the debug log but I don't see anything failing, I even get historical stats. Thank you for all the trouble you've gone through on this. 

```

2024-06-25 09:22:54.797 WARNING (MainThread) [homeassistant.util.loop] Detected blocking call to open inside the event loop by custom integration 'eyeonwater' at custom_components/eyeonwater/coordinator.py, line 43: await meter.read_historical_data( (offender: /usr/local/lib/python3.12/site-packages/pytz/__init__.py, line 108: return open(filename, 'rb')), please report it to the author of the 'eyeonwater' custom integration
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 223, in <module>
    sys.exit(main())
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 209, in main
    exit_code = runner.run(runtime_conf)
  File "/usr/src/homeassistant/homeassistant/runner.py", line 190, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 672, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 639, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1988, in _run_once
    handle._run()
  File "/usr/local/lib/python3.12/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 288, in async_refresh
    await self._async_refresh(log_failures=True)
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 312, in _async_refresh
    self.data = await self._async_update_data()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 268, in _async_update_data
    return await self.update_method()
  File "/config/custom_components/eyeonwater/__init__.py", line 52, in async_update_data
    await eye_on_water_data.read_meters(days_to_load=3)
  File "/config/custom_components/eyeonwater/coordinator.py", line 43, in read_meters
    await meter.read_historical_data(
```

**Comment by hrana** [comment]
@kdeyev Amazing! It's pulling data for me as well. Thank you again so so so much for providing the support.

@zackaryssmith I am seeing similar errors in my logs but, like yours, mine is working.

**Comment by kdeyev** [comment]
fixed in https://github.com/kdeyev/eyeonwater/releases/tag/v2.5.7

**Comment by kdeyev** [comment]
Related issues addressed in v2.6.0 (beta on branch fix/negative-statistics-issue-30). Statistics architecture reworked - uses async_add_external_statistics under eyeonwater: namespace. Breaking change: reconfigure Energy Dashboard and re-import historical data after upgrading. See #30.
